File: src/lib/parse-recipe.ts
import type { ParsedRecipe } from "./types";

function cleanText(val: unknown): string {
  if (!val) return "";
  if (Array.isArray(val)) val = val.map(String).join(" ");
  let s = String(val);
  s = s.replace(/<[^>]+>/g, "");
  // Decode common HTML entities
  s = s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, " ")
    .replace(/&#x27;/g, "'")
    .replace(/&#x2F;/g, "/");
  return s.trim();
}

function parseDuration(iso: unknown): string {
  if (!iso) return "";
  const s = String(iso);
  const m = s.match(/PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/i);
  if (!m) return s;
  let h = parseInt(m[1] || "0", 10);
  let mn = parseInt(m[2] || "0", 10);
  const sec = parseInt(m[3] || "0", 10);
  if (mn >= 60) {
    h += Math.floor(mn / 60);
    mn = mn % 60;
  }
  const parts: string[] = [];
  if (h) parts.push(`${h} hour${h !== 1 ? "s" : ""}`);
  if (mn) parts.push(`${mn} minute${mn !== 1 ? "s" : ""}`);
  if (sec && parts.length === 0) parts.push(`${sec} seconds`);
  return parts.join(" ");
}

function parseInstructions(raw: unknown): string[] {
  if (typeof raw === "string") {
    return raw
      .split(/\n+/)
      .map((s) => s.trim())
      .filter(Boolean);
  }
  if (Array.isArray(raw)) {
    const steps: string[] = [];
    for (const item of raw) {
      if (typeof item === "string") {
        steps.push(cleanText(item));
      } else if (item && typeof item === "object") {
        const obj = item as Record<string, unknown>;
        if (obj["@type"] === "HowToSection") {
          const subs = obj.itemListElement;
          if (Array.isArray(subs)) {
            for (const sub of subs) {
              if (sub && typeof sub === "object") {
                steps.push(cleanText((sub as Record<string, unknown>).text));
              }
            }
          }
        } else {
          steps.push(cleanText(obj.text));
        }
      }
    }
    return steps.filter(Boolean);
  }
  return [];
}

function parseIngredients(raw: unknown): string[] {
  if (Array.isArray(raw)) {
    return raw.map((i) => cleanText(i)).filter(Boolean);
  }
  return [];
}

function parseNutrition(raw: unknown): Record<string, string> {
  if (!raw || typeof raw !== "object") return {};
  const obj = raw as Record<string, unknown>;
  const result: Record<string, string> = {};
  const mapping: Record<string, string> = {
    calories: "calories",
    fatContent: "fat",
    carbohydrateContent: "carbs",
    proteinContent: "protein",
    fiberContent: "fiber",
    sodiumContent: "sodium",
    sugarContent: "sugar",
    cholesterolContent: "cholesterol",
  };
  for (const [key, label] of Object.entries(mapping)) {
    const val = obj[key];
    if (val) {
      result[label] = String(val).replace(" calories", "").trim();
    }
  }
  return result;
}

function getImageUrl(raw: unknown): string {
  if (typeof raw === "string") return raw;
  if (Array.isArray(raw) && raw.length > 0) return getImageUrl(raw[0]);
  if (raw && typeof raw === "object") {
    const obj = raw as Record<string, unknown>;
    return String(obj.url || obj.contentUrl || "");
  }
  return "";
}

function parseAuthor(raw: unknown): string {
  if (Array.isArray(raw)) {
    return raw
      .map((a) => parseAuthor(a))
      .filter(Boolean)
      .join(", ");
  }
  if (raw && typeof raw === "object") {
    return cleanText((raw as Record<string, unknown>).name);
  }
  return cleanText(raw);
}

export function slugify(text: string): string {
  return text
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, "")
    .replace(/[\s_]+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

function extractJsonLd(html: string): Record<string, unknown>[] {
  const pattern =
    /<script[^>]+type=["']application\/ld\+json["'][^>]*>(.*?)<\/script>/gis;
  const recipes: Record<string, unknown>[] = [];
  let match;
  while ((match = pattern.exec(html)) !== null) {
    let data: unknown;
    try {
      data = JSON.parse(match[1]);
    } catch {
      continue;
    }
    if (
      data &&
      typeof data === "object" &&
      !Array.isArray(data) &&
      "@graph" in (data as Record<string, unknown>)
    ) {
      data = (data as Record<string, unknown>)["@graph"];
    }
    const items = Array.isArray(data) ? data : [data];
    for (const item of items) {
      if (!item || typeof item !== "object") continue;
      const obj = item as Record<string, unknown>;
      const itemType = obj["@type"];
      if (Array.isArray(itemType)) {
        if (itemType.includes("Recipe")) recipes.push(obj);
      } else if (itemType === "Recipe") {
        recipes.push(obj);
      }
    }
  }
  return recipes;
}

export async function parseRecipeFromUrl(url: string): Promise<ParsedRecipe> {
  const resp = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    },
  });
  if (!resp.ok) {
    throw new Error(`Failed to fetch URL: ${resp.status}`);
  }
  const html = await resp.text();
  const recipes = extractJsonLd(html);
  if (recipes.length === 0) {
    throw new Error("No JSON-LD Recipe found on page");
  }

  const recipe = recipes[0];
  const title = cleanText(recipe.name);
  const slug = slugify(title);
  const nutrition = parseNutrition(recipe.nutrition);

  let calories: number | null = null;
  if (nutrition.calories) {
    const m = nutrition.calories.match(/\d+/);
    if (m) calories = parseInt(m[0], 10);
  }

  let tags: string[] = [];
  const keywords = recipe.keywords;
  if (typeof keywords === "string") {
    tags = keywords
      .split(",")
      .map((t) => cleanText(t))
      .filter(Boolean);
  } else if (Array.isArray(keywords)) {
    tags = keywords.map((t) => cleanText(t)).filter(Boolean);
  }

  return {
    title,
    slug,
    source_url: url,
    image_url: getImageUrl(recipe.image),
    author: parseAuthor(recipe.author),
    description: cleanText(recipe.description),
    prep_time: parseDuration(recipe.prepTime),
    cook_time: parseDuration(recipe.cookTime),
    total_time: parseDuration(recipe.totalTime),
    servings: cleanText(recipe.recipeYield),
    calories,
    ingredients: parseIngredients(recipe.recipeIngredient),
    instructions: parseInstructions(recipe.recipeInstructions),
    nutrition,
    tags,
  };
}

File: parse_recipe.py
#!/usr/bin/env python3
"""Extract JSON-LD Recipe data from a URL and output structured JSON.

Usage: python3 parse_recipe.py <url>

No external dependencies — uses only Python 3 stdlib.
"""

import json
import os
import re
import sys
import urllib.request
import urllib.error
from html import unescape
from pathlib import Path


def fetch_html(url):
    req = urllib.request.Request(url, headers={
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                      "AppleWebKit/537.36 (KHTML, like Gecko) "
                      "Chrome/120.0.0.0 Safari/537.36",
    })
    with urllib.request.urlopen(req, timeout=15) as resp:
        return resp.read().decode("utf-8", errors="replace")


def extract_json_ld(html):
    """Find all <script type="application/ld+json"> blocks and return Recipe objects."""
    pattern = r'<script[^>]+type=["\']application/ld\+json["\'][^>]*>(.*?)</script>'
    blocks = re.findall(pattern, html, re.DOTALL | re.IGNORECASE)

    recipes = []
    for block in blocks:
        try:
            data = json.loads(block)
        except json.JSONDecodeError:
            continue

        # Handle @graph arrays
        if isinstance(data, dict) and "@graph" in data:
            data = data["@graph"]

        items = data if isinstance(data, list) else [data]
        for item in items:
            if not isinstance(item, dict):
                continue
            item_type = item.get("@type", "")
            if isinstance(item_type, list):
                if "Recipe" in item_type:
                    recipes.append(item)
            elif item_type == "Recipe":
                recipes.append(item)

    return recipes


def clean_text(val):
    """Strip HTML tags and decode entities."""
    if not val:
        return ""
    if isinstance(val, list):
        val = " ".join(str(v) for v in val)
    val = str(val)
    val = re.sub(r"<[^>]+>", "", val)
    return unescape(val).strip()


def parse_duration(iso):
    """Convert ISO 8601 duration (PT1H30M) to human-readable string."""
    if not iso:
        return ""
    m = re.match(r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?", str(iso), re.IGNORECASE)
    if not m:
        return str(iso)
    hours, mins, secs = m.groups()
    h = int(hours or 0)
    mn = int(mins or 0)
    s = int(secs or 0)
    # Convert large minute values
    if mn >= 60:
        h += mn // 60
        mn = mn % 60
    parts = []
    if h:
        parts.append(f"{h} hour{'s' if h != 1 else ''}")
    if mn:
        parts.append(f"{mn} minute{'s' if mn != 1 else ''}")
    if s and not parts:
        parts.append(f"{s} seconds")
    return " ".join(parts)


def parse_instructions(raw):
    """Normalize instructions into a list of step strings."""
    if isinstance(raw, str):
        return [s.strip() for s in re.split(r'\n+', raw) if s.strip()]
    if isinstance(raw, list):
        steps = []
        for item in raw:
            if isinstance(item, str):
                steps.append(clean_text(item))
            elif isinstance(item, dict):
                if item.get("@type") == "HowToSection":
                    for sub in item.get("itemListElement", []):
                        steps.append(clean_text(sub.get("text", "")))
                else:
                    steps.append(clean_text(item.get("text", "")))
        return [s for s in steps if s]
    return []


def parse_ingredients(raw):
    if isinstance(raw, list):
        return [clean_text(i) for i in raw if clean_text(i)]
    return []


def parse_nutrition(raw):
    if not isinstance(raw, dict):
        return {}
    result = {}
    mapping = {
        "calories": "calories",
        "fatContent": "fat",
        "carbohydrateContent": "carbs",
        "proteinContent": "protein",
        "fiberContent": "fiber",
        "sodiumContent": "sodium",
        "sugarContent": "sugar",
        "cholesterolContent": "cholesterol",
    }
    for key, label in mapping.items():
        val = raw.get(key, "")
        if val:
            val = str(val).replace(" calories", "").strip()
            result[label] = val
    return result


def get_image_url(raw):
    if isinstance(raw, str):
        return raw
    if isinstance(raw, list) and raw:
        return get_image_url(raw[0])
    if isinstance(raw, dict):
        return raw.get("url", raw.get("contentUrl", ""))
    return ""


def _parse_author(raw):
    if isinstance(raw, list):
        names = [_parse_author(a) for a in raw]
        return ", ".join(n for n in names if n)
    if isinstance(raw, dict):
        return clean_text(raw.get("name", ""))
    return clean_text(raw)


def slugify(text):
    text = text.lower().strip()
    text = re.sub(r"[^\w\s-]", "", text)
    text = re.sub(r"[\s_]+", "-", text)
    text = re.sub(r"-+", "-", text)
    return text.strip("-")


def download_image(url, slug):
    """Download image and save to images/ directory. Returns relative path."""
    if not url:
        return ""
    ext = ".jpg"
    m = re.search(r"\.(jpe?g|png|webp|gif)", url, re.IGNORECASE)
    if m:
        ext = "." + m.group(1).lower()
        if ext == ".jpeg":
            ext = ".jpg"

    images_dir = Path(__file__).parent / "images"
    images_dir.mkdir(exist_ok=True)
    filename = f"{slug}{ext}"
    filepath = images_dir / filename

    try:
        req = urllib.request.Request(url, headers={
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                          "AppleWebKit/537.36 (KHTML, like Gecko) "
                          "Chrome/120.0.0.0 Safari/537.36",
        })
        with urllib.request.urlopen(req, timeout=15) as resp:
            filepath.write_bytes(resp.read())
        return f"/images/{filename}"
    except Exception as e:
        print(f"Warning: could not download image: {e}", file=sys.stderr)
        return ""


def parse_recipe(url):
    html = fetch_html(url)
    recipes = extract_json_ld(html)

    if not recipes:
        print("Error: no JSON-LD Recipe found on page.", file=sys.stderr)
        sys.exit(1)

    recipe = recipes[0]
    title = clean_text(recipe.get("name", ""))
    slug = slugify(title)

    image_url = get_image_url(recipe.get("image", ""))
    local_image = download_image(image_url, slug)

    nutrition = parse_nutrition(recipe.get("nutrition", {}))
    calories_num = ""
    if nutrition.get("calories"):
        m = re.search(r"\d+", nutrition["calories"])
        if m:
            calories_num = int(m.group())

    result = {
        "title": title,
        "slug": slug,
        "source_url": url,
        "image": local_image,
        "image_url": image_url,
        "author": _parse_author(recipe.get("author", "")),
        "description": clean_text(recipe.get("description", "")),
        "prep_time": parse_duration(recipe.get("prepTime", "")),
        "cook_time": parse_duration(recipe.get("cookTime", "")),
        "total_time": parse_duration(recipe.get("totalTime", "")),
        "servings": clean_text(recipe.get("recipeYield", "")),
        "calories": calories_num,
        "ingredients": parse_ingredients(recipe.get("recipeIngredient", [])),
        "instructions": parse_instructions(recipe.get("recipeInstructions", [])),
        "nutrition": nutrition,
        "tags": [clean_text(t) for t in recipe.get("keywords", "").split(",") if clean_text(t)] if isinstance(recipe.get("keywords"), str) else [clean_text(t) for t in recipe.get("keywords", []) if clean_text(t)],
    }

    return result


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <url>", file=sys.stderr)
        sys.exit(1)

    data = parse_recipe(sys.argv[1])
    print(json.dumps(data, indent=2))
